infinispan-embedded/src/main/java/org/hibernate/ogm/datastore/infinispan/persistencestrategy/counter/SequenceClusteredCounterHandler.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.datastore.infinispan.persistencestrategy.counter;
+
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.hibernate.HibernateException;
+import org.hibernate.boot.model.relational.Sequence;
+import org.hibernate.ogm.datastore.infinispan.logging.impl.Log;
+import org.hibernate.ogm.datastore.infinispan.logging.impl.LoggerFactory;
+import org.hibernate.ogm.dialect.spi.NextValueRequest;
+import org.infinispan.counter.EmbeddedCounterManagerFactory;
+import org.infinispan.counter.api.CounterManager;
+import org.infinispan.counter.api.StrongCounter;
+import org.infinispan.manager.EmbeddedCacheManager;
+
+/**
+ * Provides access to Infinispan Clustered Counter feature.
+ * Used by the dialect to implement a reliable ID generator.
+ *
+ * The class is used when the strategy type of ID generator is SequenceGenerator.
+ * The required and not declared counters are created by the dialect at start up.
+ *
+ * @author Fabio Massimo Ercoli
+ */
+public class SequenceClusteredCounterHandler extends ClusteredCounterHandler {
+
+	private static final Log LOG = LoggerFactory.make( MethodHandles.lookup() );
+
+	// true if the current instance has created the counter
+	// and the counter is never used
+	private final AtomicBoolean returnTheInitialValue;
+
+	private final StrongCounter counter;
+
+	public SequenceClusteredCounterHandler(EmbeddedCacheManager cacheManager, Sequence sequence) {
+
+		super( cacheManager );
+
+		CounterManager counterManager = EmbeddedCounterManagerFactory.asCounterManager( cacheManager );
+
+		returnTheInitialValue = new AtomicBoolean(
+			defineCounterIfNotExists( sequence.getInitialValue(), counterManager, sequence.getExportIdentifier() )
```

Comment:
The issue here is to return the initial value of the counter at the first use of it and applying the default behaviour for all other interactions. If the counter is new, the next and only the next request of nextValue() should return the initial value, instead of increment and return new values. According to the tests: SequenceNextValueGenerationTest and TableNextValueGenerationTest. If we could skip the initial value, it would all be much simpler :P